graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java

```diff
@@ -286,6 +297,204 @@ public Object absObject(VirtualFrame frame, Object object,
+        }
+
+        @Specialization(limit = "3")
+        protected boolean doHashStorage(VirtualFrame frame,
+                        HashingStorage hashingStorage,
+                        NodeType nodeType,
+                        @CachedLibrary("hashingStorage") HashingStorageLibrary hlib) {
+            for (Object key : hlib.keys(hashingStorage)) {
+                LoopNode.reportLoopCount(this, 1);
+                if (nodeType == NodeType.ALL) {
+                    if (!isTrueNode.execute(frame, key)) {
+                        return false;
+                    }
+                } else if (nodeType == NodeType.ANY && isTrueNode.execute(frame, key)) {
+                    return true;
+                }
+            }
+
+            return nodeType == NodeType.ALL;
+        }
+    }
+
+    @Builtin(name = ALL, minNumOfPositionalArgs = 1)
+    @GenerateNodeFactory
+    public abstract static class AllNode extends PythonUnaryBuiltinNode {
+
+        @Specialization(guards = "cannotBeOverridden(object, getClassNode)", limit = "1")
+        static boolean doList(VirtualFrame frame,
+                        PList object,
+                        @SuppressWarnings("unused") @Shared("getClassNode") @Cached GetClassNode getClassNode,
+                        @Shared("allOrAnyNode") @Cached AllOrAnyNode allOrAnyNode) {
+            return allOrAnyNode.execute(frame, object.getSequenceStorage(), AllOrAnyNode.NodeType.ALL);
+        }
+
+        @Specialization(guards = "cannotBeOverridden(object, getClassNode)", limit = "1")
+        static boolean doTuple(VirtualFrame frame,
+                        PTuple object,
+                        @SuppressWarnings("unused") @Shared("getClassNode") @Cached GetClassNode getClassNode,
+                        @Shared("allOrAnyNode") @Cached AllOrAnyNode allOrAnyNode) {
+            return allOrAnyNode.execute(frame, object.getSequenceStorage(), AllOrAnyNode.NodeType.ALL);
+        }
+
+        @Specialization(guards = "cannotBeOverridden(object, getClassNode)", limit = "1")
+        static boolean doHashColl(VirtualFrame frame,
+                        PHashingCollection object,
+                        @SuppressWarnings("unused") @Shared("getClassNode") @Cached GetClassNode getClassNode,
+                        @Shared("allOrAnyNode") @Cached AllOrAnyNode allOrAnyNode) {
+            return allOrAnyNode.execute(frame, object.getDictStorage(), AllOrAnyNode.NodeType.ALL);
+        }
+
+        @Specialization
+        boolean doObject(VirtualFrame frame,
+                        Object object,
+                        @Cached PyObjectGetIter getIter,
+                        @Cached GetNextNode nextNode,
+                        @Cached IsBuiltinClassProfile errorProfile,
+                        @Cached PyObjectIsTrueNode isTrueNode) {
+            Object iterator = getIter.execute(frame, object);
+            while (true) {
+                try {
+                    LoopNode.reportLoopCount(this, 1);
```

Comment:
The reporting is supposed to be used to report the total number of iterations after the loop has finished. I'm not sure what happes when you report multiple times in a loop like this, but I think the compiler would simply average the reported numbers, so you would end up with the compiler assuming the loop is always executed once. What I believe you should do instead is to just count the number of iterations in a local variable and report it in a `finally` block as the last thing.

Reply:
Changed in bceb24a ; my bad for not reading the documentation closely enough for this one, from examples in the code I'd seen elsewhere, I had been assuming it aggregated the values fed for each `reportLoopCount` call in each scope (and I was really curious as to how that automatic detection worked...) 

Also, the infinite loops and better profiling for the (ex) for-each loop were fixed in 522e6ed.